### PR TITLE
Adjustable master avatar volume

### PR DESCRIPTION
**Pull request by kencooke**
This PR adds an avatar "master volume" slider to Interface, which tells the server what level to mix all avatar voices for this client. This is in addition to the per-avatar levels in PAL, and applies to all current/future avatars in the domain. Like PAL, it is not persistent across domains or Interface sessions.

A large gain (eg +20dB) has the effect that all avatars sound much louder, and when close to you, will activate dynamic range compression to prevent clipping. Sound effects (injectors) and Ambisonic environments are not affected by this setting, so it can also be used to control the mix.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4571/HighFidelity-Beta-PR11718-cfba6ae819cbcdb64dcbd5583dbe0334ed09e6c6.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4571/HighFidelity-Beta-PR11718-cfba6ae819cbcdb64dcbd5583dbe0334ed09e6c6.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4596/HighFidelity-Beta-PR11718-82d9827571d0b8fb2e304267176754704c2beb95.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4596/HighFidelity-Beta-PR11718-82d9827571d0b8fb2e304267176754704c2beb95.exe

**Comment by kencooke** [comment]
Jenkins, test this please

**Comment by kencooke** [comment]
### Test Plan

1) Install the client and server from this PR, and go to your local domain. Have at least one other avatar (running the existing Interface client) in the domain. Verify you can adjust the level of the other avatars, using the new "master gain" slider in PAL (see screenshot below). Note that your own audio (via server echo) is not changed by this slider.
In your Interface log, you should see entries such as
```"[11/01 08:56:58] [DEBUG] [hifi.networking] Sending Set MASTER Avatar Gain packet with Gain: 5"```
and in the audio-mixer log,
```[11/01 08:06:02] [DEBUG] [default] [13132] [audio-mixer] Setting MASTER avatar gain for  QUuid("{be66d7c3-b9e6-4de2-8ab1-b3c1b2532111}")  to  1.77828```

2) When you join an existing domain with Interface from this PR, audio should work correctly and the "master gain" slider will be ignored.

3) Verify that audio otherwise works as expected.


untitled

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4597/HighFidelity-Beta-PR11718-82d9827571d0b8fb2e304267176754704c2beb95.dmg

**Comment by Nex-Pro** [comment]
Awaiting win build

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4597/HighFidelity-Beta-PR11718-82d9827571d0b8fb2e304267176754704c2beb95.exe

**Comment by Nex-Pro** [comment]
Testing

**Comment by Nex-Pro** [comment]
### PASS

When this interface is used at dev-welcome the slider is ignored.

When used at the domain server from this pr other users ( not running this interface build ) are able to connect and their volume is correctly adjusted. No adverse effect on sound quality noticed.

**Comment by kencooke** [comment]
Adding my child-like diagram of the mixing bus, mostly for comedic value.

audio-mixer

**Comment by themelissabrown** [comment]
Let's check in with Design on what the name of it should be, but the code works really well! Thanks Ken!

**Comment by RyanDowne** [comment]
How difficult would it be to move this master volume slider out of the PAL and into something else? For example, the audio settings dialog? Asking for a friend.

**Comment by kencooke** [comment]
@RyanDowne It's a bit of work to re-implement it somewhere else but a QML expert could surely do it. I'd recommend a future PR so we can get feedback on any changes to the functionality also.

I integrated it with PAL for two reasons -- it recycles the existing PAL code, and the behavior is related to the existing PAL gains (both are avatar-audio only, interact with each other, and not persistent across domains/sessions). Moving it to the Audio tab might seem like it adjusts your audio device volume, which is something quite different.

File: assignment-client/src/audio/AudioMixerClientData.cpp [hunk 1]
```diff
@@ -189,8 +189,16 @@ void AudioMixerClientData::parsePerAvatarGainSet(ReceivedMessage& message, const
...
+
+    if (avatarUuid.isNull()) {
+        // set the MASTER avatar gain
+        setMasterAvatarGain(gain);
+        qDebug() << "Setting MASTER avatar gain for " << uuid << " to " << gain;
+    } else {
+        // set the per-source avatar gain
+        hrtfForStream(avatarUuid, QUuid()).setGainAdjustment(gain);
+        qDebug() << "Setting avatar gain adjustment for hrtf[" << uuid << "][" << avatarUuid << "] to " << gain;
+    }
```
Comment (AndrewMeadows) [inline assignment-client/src/audio/AudioMixerClientData.cpp]: Maybe this should be using the `qCDebug()` API which allows for a "category" argument?  It looks like there are some "audio" logging categories declared:
```
./libraries/audio/src/AudioLogging.h:Q_DECLARE_LOGGING_CATEGORY(audio)
./libraries/audio/src/AudioLogging.h:Q_DECLARE_LOGGING_CATEGORY(audiostream)
```

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4661/HighFidelity-Beta-PR11718-68e9ec9b23862e00295c4dbe084a09cd1a44449d.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4661/HighFidelity-Beta-PR11718-68e9ec9b23862e00295c4dbe084a09cd1a44449d.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4755/HighFidelity-Beta-PR11718-addcb51ed98e09693f79dd6c8bcf3f063b1f58df.exe

**Comment by ZappoMan** [comment]
Is there a reason this is DO NOT MERGE? @RyanDowne - what PM owns this, is this RC59 targetted, etc etc

File: assignment-client/src/audio/AudioMixerSlave.cpp [hunk 2]
```diff
@@ -484,10 +484,12 @@ float approximateGain(const AvatarAudioStream& listeningNodeStream, const Positi
     // when throttling, as close streams are expected to be heard by a user
     float distance = glm::length(relativePosition);
     return gain / distance;
+
+    // avatar: skip master gain - it is constant for all streams
 }
 
-float computeGain(const AvatarAudioStream& listeningNodeStream, const PositionalAudioStream& streamToAdd,
-        const glm::vec3& relativePosition, bool isEcho) {
+float computeGain(AudioMixerClientData& listenerNodeData, const AvatarAudioStream& listeningNodeStream,
```
Comment (hyperlogic) [inline assignment-client/src/audio/AudioMixerSlave.cpp]: Can you make listenerNodeData can be a const reference?  AudioMixerClientData::getMasterAvatarGain() is a const method, so it should work.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4935/HighFidelity-Beta-PR11718-9810d1582d6d48f8fe7ff7e44c4e8eb4b6199f7d.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4935/HighFidelity-Beta-PR11718-9810d1582d6d48f8fe7ff7e44c4e8eb4b6199f7d.exe

File: assignment-client/src/audio/AudioMixerSlave.cpp (hunk 2 above)
Comment (kencooke) [reply assignment-client/src/audio/AudioMixerSlave.cpp]: Good idea, done.

**Pull request merged**